### ISSUE TITLE
Dynamic Schema

### ISSUE THREAD

**Issue by YannickB** [overlookmotel/sequelize-hierarchy#129]
Hello all,

I'm trying to make sequelize-hierarchy work in my application with multiple schemas, but I struggle to make it work. My application have several schemas named based on the url of the application, all with same tables.

My object Requirement has hierarchy

```
const Requirement = Model.define('requirement', {

  title: {
    allowNull: false,
    type: DataType.STRING
  },
  parentId: {
    type: DataType.INTEGER,
    references: {
      model: 'requirements',
      key: 'id'
    },
    field: 'parent_id',
    hierarchy: {
      through: 'ancestor',
      throughTable: 'requirement_ancestors'
    }
  },
...

```

I can't specify here the schema with `throughSchema` because the schema is dynamic and change depending the url (hence depending a parameter in request, unaccessible here). I precise that tables requirements and requirement_ancestors are correctly created.

Here is how I create a requirement : 
```
Requirement.schema(context.schema).create(vals);
```

The `schema()` function http://docs.sequelizejs.com/class/lib/model.js~Model.html#static-method-schema add the schema in my requirement insert.
```
Executing (default): INSERT INTO "my.localhost"."requirements" ("id","title","parent_id","sequence","computation","weight","state","can_have_children","created_at","updated_at","hierarchy_level") VALUES (DEFAULT,'Requirement 2.1.1',36,1,'parent-weighted-average',2,'published',true,'2017-07-20 21:31:15.443 +00:00','2017-07-20 21:31:15.443 +00:00',2) RETURNING *;
```

But it fail to use the schema in the ancestors table
```
Executing (default): INSERT INTO "requirement_ancestors" ("requirement_id","ancestor_id") VALUES (37,36);
relation "requirement_ancestors" does not exist
```

Is there a way to specify the schema ancestors have to use in my create call ?

**Comment by overlookmotel** [comment]
Hmm. I don't use Postgres so no experience of schemas. However, sequelize-hierarchy should support them (it was added in a PR and there are tests covering their use).

Presumably it'd make sense for the hierarchy table to always be in same schema as the main table. So if you create the same table in multiple schemas, it'd also create the hierarchy table in multiple schemas. And then any query on a table in a certain schema should use the hierarchy table in that same schema?

What I mean is that there'd be no need to dynamically specify the schema ancestors have to use in the create call - it'd be automatically using the schema you'd expect.

Does that logic make sense to you?

I thought this was what was already happening and the tests do run using schemas so I thought this was covered by the tests. However, perhaps the tests are running on the ancestor table in default schema created *in a previous test*.

I've altered the tests so it drops all tables after each test. That should reveal the problem if it's there. Tests are running now on Travis: https://travis-ci.org/overlookmotel/sequelize-hierarchy/builds/256131349

Let's see what happens!

**Comment by overlookmotel** [comment]
Argh tests are failing for Postgres with schemas. But not in the place I expected they might. Will look into it. But let me know if the logic I'm suggesting makes sense to you.

**Comment by YannickB** [comment]
Hi @overlookmotel,

Yes I agree with you, the hierarchy table should always be in the same schema than the main table, hence there is no need for specifying the schema for the hierarchy table. 
At least this is not my use case, the hierarchy table is in the same schema, when I named the issue "Dynamic schema" if was only for taking into account the .schema() function I used for the main table.

It looks indeed like a bug, let's see what the tests will reveal. Thanks for your help !

**Comment by overlookmotel** [comment]
The tests all pass. But yes it is a bug.

I don't use Postgres and so don't really understand schemas. Schema support was added in a PR by @teagup (https://github.com/overlookmotel/sequelize-hierarchy/pull/41) and then some further commits in response to issues https://github.com/overlookmotel/sequelize-hierarchy/issues/46 and https://github.com/overlookmotel/sequelize-hierarchy/issues/60.

Here's how it works at present:

* When a model is defined, the hierarchy table is created in same schema as the main table
* The hierarchy model is recorded on the main model as `Model.hierarchy.through`
* When a method (e.g. `.create()`) is called on a model, the hooks code in sequelize-hierarchy then makes all changes to hierarchy table via method calls on `model.hierarchy.through`

In your case, using the `.schema()` function I guess is not returning the right model, or rather it's just returning the no-schema `Requirement` model with some extra attributes set. So `model.hierarchy.through` still refers to the no-schema hierarchy table. Hence the problem you're observing.

What I don't understand about working with schemas is whether you are calling `.define()` multiple times to create the tables in each schema, or whether it works some other way.

If the former, in my opinion, Sequelize's `.schema()` should return the version of the model attached to the schema you've requested. And then sequelize-hierarchy would do its thing quite correctly. So then I'd say this is caused by a bug in Sequelize itself.

But if it works some other way from this, then I don't understand what the situation is.

But what I'm pretty sure of is that there should not need to be a dynamic `schema` option for calling `.create()` or any other methods. One base table should be locked to one hierarchy table. Updating one updates the other. So you should never need to specify the schema of the hierarchy table when calling `.create()` - it should *always* be the table that was created as the hierarchy table when the model was defined. So I would prefer not to add a dynamic schema option as a workaround for the issue you've encountered - it would be better to solve the root cause of the problem.

I'm afraid at this point I'm not sure I can be much more help. It's hard for me to make headway on this without installing Postgres, working out how to use it, and diving into schemas to understand how they work. As I have no other use for Postgres myself, I'm afraid this is time I can't afford to devote. Sorry if that sounds bad - just obviously we all have finite time and day jobs to get on with! I'm just trying to be realistic.

However... if you'd like to try to solve the problem and submit a PR I'd be very happy to receive it. Or indeed, if the problem is with Sequelize itself as I mention above, it'd be great if you could submit a PR on Sequelize. And I'm very happy to answer questions about the internals of this module to point you in the right direction / clarify anything you don't understand (just don't ask me about schemas!)

Also @teagup might you be able to assist with this at all?

**Comment by YannickB** [comment]
No worries, I already saw before creating the issue that you didn't implemented the schema management by yourself so I was kinda expecting this answer. My main concern was making it recognized as a bug and I think this job is done.

Regarding the table creation, I use sequelize migration to manually create main table and hierarchy table, not .sync() so I had no problem here (umzug was really handy to manage the schema creation programmabtically without using sequelize-cli)

I agree we need to fix the root cause, no need to add an option on create. Also note that here I use .schema() function but in sequelize we can also use the searchPath option to recognize the schema. I decided not to use it because I find .schema() more readable.

On my side I don't really know how to set up a development environnment to contribute the module, but I guess I can just make the change in the node_modules folder. Fixing this is still the best way for me to handle hierarchy so you can be sure I'll do my best sooner or later, it'll just take me time to understand how the depth works. Of course indeed help and/or opinion from the contributor who added schema support will be greatly appreciated.

Again thanks for your time and the deep analysis.

**Comment by overlookmotel** [comment]
@YannickB No problem. Thanks for finding the bug and raising this issue. If you are able to find a fix, that'd be great. If you don't have time yourself either, I understand!

I've found the easiest way to contribute fixes to open source code on Github is:
1. Fork the repo (fork button at top right of this page)
2. Download and install Github Desktop
3. On your fork, click the "Open in Github Desktop" button
4. In Terminal, navigate to the folder you downloaded the code to
5. Install dependencies with `npm install`

Then you can run the tests with `npm test`.

You can then make changes, commit them to your "fork" of the repo and submit them as a pull request. Might take a little time the first time to figure out the process, but once you've done it once, it's easy easy easy.

**Comment by YannickB** [comment]
Ok, I believe the correct way to retrieve the schema is by using `.getTableName()` http://docs.sequelizejs.com/class/lib/model.js~Model.html#static-method-getTableName, this shall return an object with the schema.

I tried to call the function in https://github.com/overlookmotel/sequelize-hierarchy/blob/master/lib/hooksModel.js#L46 but this doesn't seems to work. Also I'm not sure where is the best place to insert the schema, any hint here would be welcome.

I continue looking deeper into the problem

**Comment by overlookmotel** [comment]
Hi @YannickB.

I've looked at your PR #130 but replying here to keep discussion in one place.

Thanks loads for looking into this and attempting a fix, but I'm afraid I think you're taking the wrong approach.

The schema that the hierarchy table is in is not necessarily the same as the schema the main table is in. That's the default behaviour, but it can be overruled by `throughSchema` option when defining the model.

The through model should *always* be the model referred to by the property `.hierarchy.through` on the main model.

The point is that calling `.schema()` does not return the correct model in some way. If it returned the right model then `.hierarchy.through` would point to the right hierarchy table in the right schema.

So I'm sorry to say I think that your PR is trying the fix the problem in the wrong place.

Again, I don't fully understand schemas, so I may have this wrong, but the above is my best guess.

**Comment by YannickB** [comment]
Hi @overlookmotel 

I really don't know about it... `.hierarchy.through` seems static to me in every case, it is set directly after the `define()` function. Then when we call for example a `MyModel.schema(myschema).create(vals)` the `.hierarchy.through` is already defined.

That's why I think our only option is to override the hooks create and update by using there the `getTableName()` which is the only way to correctly get the schema provided by the `schema()` function. At least that's what I believe, please prove me wrong.

But the create and update function are not the issue anymore. As I said, `getTableName()` works, now my issue is to correctly create the hierarchy table inside the schema with `.sync()` and here I'm stuck because I don't know how I can access to the schema provided as option of `.sync()`, which seems the obvious way to deploy the tables inside a specified schema.

**Comment by overlookmotel** [comment]
I don't use Postgres so this schema business is a bit beyond me. Is anyone interested in trying to implement this? If not, I'm going to close this issue.

**Comment by overlookmotel** [comment]
Well, no-one has piped up since April, so I'm going to give up on this and close 2 incomplete PRs #144 and #130 which are related.

If someone really wants this fixed, I will be happy to accept a PR. However, to be clear, I do not use Postgres myself and the entire concept of schemas seems pretty weird to me. I don't get it, and it's of no benefit to me to spend the time to get it. So, if someone wants to tackle it, I'd ask them to:

1. Develop a complete solution that's not a stopgap
2. Implement it, write the tests, document it - the whole shebang
3. Explain to me enough about schemas that I can validate that the solution you've come up with is valid.

Basically, you'll need to do all the running.

Sorry if this sound harsh, but I'm a bit burned by the original implementation being flawed, and those who implemented it disappearing leaving me to try to maintain a mess which I don't understand.

It's all totally understandable how that happened - all part of the rough and tumble of open source, and no disrespect to those who submitted the flawed implementation - but I'm not going to make the same mistake twice!
